File: scripts/dev_proxy_2026.mjs
import http from "node:http";

const port = 2026;

function resolveTarget(pathname) {
  if (pathname.startsWith("/api/langgraph/")) {
    return {
      hostname: "127.0.0.1",
      port: 2024,
      path: pathname.slice("/api/langgraph".length) || "/",
    };
  }

  if (
    pathname.startsWith("/api/") ||
    pathname === "/health" ||
    pathname === "/docs" ||
    pathname === "/redoc" ||
    pathname === "/openapi.json"
  ) {
    return {
      hostname: "127.0.0.1",
      port: 8001,
      path: pathname,
    };
  }

  return {
    hostname: "127.0.0.1",
    port: 3000,
    path: pathname,
  };
}

const server = http.createServer((req, res) => {
  const { hostname, port: targetPort, path } = resolveTarget(req.url || "/");
  const upstream = http.request(
    {
      hostname,
      port: targetPort,
      path,
      method: req.method,
      headers: {
        ...req.headers,
        host: `${hostname}:${targetPort}`,
        connection: "close",
      },
    },
    (upstreamRes) => {
      const headers = { ...upstreamRes.headers };
      delete headers["content-encoding"];
      delete headers["transfer-encoding"];
      headers.connection = "close";
      res.writeHead(upstreamRes.statusCode || 502, headers);
      upstreamRes.pipe(res);
    },
  );

  upstream.on("error", (error) => {
    const message = Buffer.from(String(error));
    res.writeHead(502, {
      "content-type": "text/plain; charset=utf-8",
      "content-length": String(message.length),
      connection: "close",
    });
    res.end(message);
  });

  req.pipe(upstream);
});

server.listen(port, "127.0.0.1", () => {
  console.log(`proxy listening on http://127.0.0.1:${port}`);
});

File: scripts/dev_proxy_2026.py
import http.server
import socketserver
import sys
import urllib.error
import urllib.request


class ProxyHandler(http.server.BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.0"

    def _target(self) -> str:
        path = self.path
        if path.startswith("/api/langgraph/"):
            return "http://127.0.0.1:2024/" + path[len("/api/langgraph/") :]
        if path.startswith("/api/") or path in {"/health", "/docs", "/redoc", "/openapi.json"}:
            return "http://127.0.0.1:8001" + path
        return "http://127.0.0.1:3000" + path

    def _proxy(self) -> None:
        target = self._target()
        body = None
        length = self.headers.get("Content-Length")
        if length:
            body = self.rfile.read(int(length))

        request = urllib.request.Request(target, data=body, method=self.command)
        for key, value in self.headers.items():
            if key.lower() in {"host", "connection", "content-length", "accept-encoding"}:
                continue
            request.add_header(key, value)

        try:
            with urllib.request.urlopen(request, timeout=600) as response:
                payload = response.read()
                self.send_response(response.status)
                for key, value in response.getheaders():
                    if key.lower() in {"transfer-encoding", "connection", "content-encoding"}:
                        continue
                    self.send_header(key, value)
                self.send_header("Connection", "close")
                self.send_header("Content-Length", str(len(payload)))
                self.end_headers()
                if payload:
                    self.wfile.write(payload)
        except urllib.error.HTTPError as error:
            payload = error.read()
            self.send_response(error.code)
            for key, value in error.headers.items():
                if key.lower() in {"transfer-encoding", "connection", "content-encoding"}:
                    continue
                self.send_header(key, value)
            self.send_header("Connection", "close")
            self.send_header("Content-Length", str(len(payload)))
            self.end_headers()
            if payload:
                self.wfile.write(payload)
        except Exception as error:
            payload = str(error).encode("utf-8")
            self.send_response(502)
            self.send_header("Content-Type", "text/plain; charset=utf-8")
            self.send_header("Connection", "close")
            self.send_header("Content-Length", str(len(payload)))
            self.end_headers()
            self.wfile.write(payload)

    def do_GET(self) -> None:
        self._proxy()

    def do_POST(self) -> None:
        self._proxy()

    def do_PUT(self) -> None:
        self._proxy()

    def do_PATCH(self) -> None:
        self._proxy()

    def do_DELETE(self) -> None:
        self._proxy()

    def do_OPTIONS(self) -> None:
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, PUT, PATCH, DELETE, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "*")
        self.send_header("Content-Length", "0")
        self.end_headers()

    def log_message(self, fmt: str, *args) -> None:
        sys.stdout.write("%s - - [%s] %s\n" % (self.client_address[0], self.log_date_time_string(), fmt % args))
        sys.stdout.flush()


class ThreadingServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
    daemon_threads = True
    allow_reuse_address = True


if __name__ == "__main__":
    server = ThreadingServer(("127.0.0.1", 2026), ProxyHandler)
    print("proxy listening on http://127.0.0.1:2026", flush=True)
    server.serve_forever()
